File: frontend/src/types/Book.ts
import type { Author } from "./Author";

export type ReadingStatus = "to_read" | "reading" | "read" | "in_library";
export type BookFormat = "paperback" | "ebook" | "audiobook";

export interface Reading {
  id: number;
  book_id: number;
  status: ReadingStatus;
  start_reading?: string;
  finish_reading?: string;
  rating?: number;
  notes?: string;
  new_author: boolean;
}

export interface Book {
  id: number;
  title: string;
  isbn?: string;
  pages?: number;
  author_id: number;
  author: Author;
  genre?: string;
  year?: number;
  language?: string;
  publisher?: string;
  category?: string;
  format?: BookFormat;
  readings: Reading[];
}

export interface BookCreate {
  title: string;
  isbn?: string;
  pages?: number;
  author_id: number;
  genre?: string;
  year?: number;
  language?: string;
  publisher?: string;
  category?: string;
  format?: BookFormat;
  start_reading?: string;
  finish_reading?: string;
  rating?: number;
  notes?: string;
  new_author?: boolean;
}

export interface ReadingCreate {
  book_id: number;
  status: ReadingStatus;
  start_reading?: string;
  finish_reading?: string;
  rating?: number;
  notes?: string;
  new_author?: boolean;
}

File: backend/app/schemas/book.py
from datetime import date

from pydantic import BaseModel, ConfigDict, Field

from app.models.enums import ReadingStatus, BookFormat

class BookBase(BaseModel):
    title: str
    isbn: str | None = None
    pages: int | None = Field(default=None, gt=0)
    author_id: int
    status: ReadingStatus
    start_reading: date | None = None
    finish_reading: date | None = None
    rating: int | None = Field(default=None, ge=1, le=5)
    notes: str | None = None
    genre: str | None = None
    year: int | None = None
    language: str | None = None
    publisher: str | None = None
    category: str | None = None
    format: BookFormat | None = None


class BookCreate(BookBase):
    pass


class BookResponse(BookBase):
    id: int

    model_config = ConfigDict(from_attributes=True)